### PR TITLE
Added basic prompt generation for RAG.

### PR DESCRIPTION
**Pull request by morgan55555**
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [x] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Have you written and run sufficient tests for validating the changes?
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Prefix:** To cleary categorize this pull request, prefix the pull request title, using one of the following:
  - **feat**: Introduces a new feature or enhancement to the codebase

# Changelog Entry

### Description

- Adding a feature: option to generate more actual RAG query by llm, then process RAG search.

### Added

- 'Enable RAG Prompt Generation' option, disabled by default.
- 'RAG Prompt Generation Template' option with default template.

### Changed

- RAG documents search is now using prompt before search, if 'Enable RAG Prompt Generation' is selected.

---

### Additional Information

- [Insert any additional context, notes, or explanations for the changes]
  - Based on idea from issue: https://github.com/open-webui/open-webui/issues/5571

### Screenshots or Videos

image

### CODE REVIEW

**Comment by jannikstdl** [comment]
Thank you for this PR! I've tested it on my device and can confirm it works as expected.

For further optimization, we could experiment with different RAG generation prompts to enhance the quality of vector search queries.

For example, I tried to reform the RAG generation prompt to:

- Guide the model to be aware of follow-up questions to generate contextually relevant vector search queries.
- Instruct the model to produce at least three variations of the vector search string by rephrasing the user query for enhanced search accuracy. This approach works to some extent; however, I think the best quality solution would be an option to create three independent vector queries and merge the top K results into a single, high-quality output.

In my testing, the model consistently responded without generating empty strings, so i could not test what this would do and when it would be useful.



Here are some additional ideas:

With the recent release of the knowledge feature, we could further develop this to support agentic RAG decisions. For instance, we could modify the task to let the model compare the knowledge description and the user query, determining if a RAG search is necessary in the first place.

This addition should fully address #5571.

**Comment by morgan55555** [comment]
> * Guide the model to be aware of follow-up questions to generate contextually relevant vector search queries.
> * Instruct the model to produce at least three variations of the vector search string by rephrasing the user query for enhanced search accuracy. This approach works to some extent; however, I think the best quality solution would be an option to create three independent vector queries and merge the top K results into a single, high-quality output.

Seems that more useful will be generating multiple search query, one query for one subject, and do RAG search for every query, so, this will be (top K * queries) results, for every subject in prompt. For example, prompt "Write Lua code that moves an object in a circle. Use Scenario template." can be divided into "How to move a object?" + "What is Scenario".

I will implement it later, after first (this) version will be merged.

File: backend/open_webui/apps/retrieval/main.py [hunk 1]
```diff
@@ -86,6 +86,8 @@
     RAG_WEB_SEARCH_DOMAIN_FILTER_LIST,
     RAG_WEB_SEARCH_ENGINE,
     RAG_WEB_SEARCH_RESULT_COUNT,
+    ENABLE_RAG_PROMPT_GENERATION,
+    RAG_PROMPT_GENERATION_TEMPLATE,
```
Comment (tjbck) [inline backend/open_webui/apps/retrieval/main.py:90]: `RAG_QUERY_GENERATION_TEMPLATE`

Comment (morgan55555) [reply backend/open_webui/apps/retrieval/main.py]: Done!

**Comment by oloflarsson** [comment]
@morgan55555 looking forward to this merge request. I feel like it's going to make a big difference 👏!

**Comment by morgan55555** [comment]
So, @tjbck, can it be merged?

**Comment by jannikstdl** [comment]
> So, @tjbck, can it be merged?

He is busy ATM integrating the core features for 0.4, but im sure this great PR will be merged soon!

**Comment by tjbck** [comment]
Cherry-picking changes, get_rag_context requires `messages` to be passed. Appreciate the effort.

**Comment by morgan55555** [comment]
So, when it will be in dev version? Do you need any help?

**Comment by tjbck** [comment]
This is already in main 😅

**Comment by jereloh** [comment]
Hi! I was looking for something like this PR, glad to see this initiative but I do not see the option presented in the new releases - I'm on 0.6.25. What happened to this?

**Comment by morgan55555** [comment]
> Hi! I was looking for something like this PR, glad to see this initiative but I do not see the option presented in the new releases - I'm on 0.6.25. What happened to this?

Same prompt used for web and RAG search.

**Pull request closed without merging**